### PR TITLE
KAFKA-7223: Add late-record metrics

### PR DESCRIPTION
**Pull request by vvcephei**
Add late record metrics, as specified in KIP-328

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java:75]: The pre-canned Max-and-Avg function in StreamsMetricsImpl is specifically for latencies, which this is not. Hence adding the metrics right here.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java]: The actual value depends on how quickly the test runs, but we do know it shouldn't be 0, which is sufficient to demonstrate that *something* got recorded.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java:76]: on-the-side, avoiding deprecated APIs in an unrelated test.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java]: silly formatting fix

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java:294]: This class has two different tests that check the metrics, so I pulled it out into a function.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java:340]: With the prior test data, all the records would have the exact same amount of lateness. I created a difference, which lets us demonstrate correct max and avg lateness metrics.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java]: I changed advance from 5 to 10, just to make the logs and output less noisy.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java:389]: no point in parameterizing this one, it would be the same each time.

File: streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java]: incidental reformatting.

File: streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java]: I didn't like matching on the exact Double value, and Hamcrest doesn't seem to have a matcher that does something like this...

**Review by vvcephei — COMMENTED** [review]
@guozhangwang @mjsax @bbejeck ,

This is the first chunk of metrics from the KIP: https://cwiki.apache.org/confluence/display/KAFKA/KIP-328%3A+Ability+to+suppress+updates+for+KTables#KIP-328:AbilitytosuppressupdatesforKTables-Metrics

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: nit: remove `since`

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java]: same as above

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: I'm not 100% sure I agree with window end equaling close time makes a record late, but I can see the point, so I'll go with what others think.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: Could this be a `DEBUG` level recording? Again no strong opinion here.

**Comment by vvcephei** [comment]
Something crazy happened with the core tests.

Retest this, please.

**Review by bbejeck — COMMENTED** [review]
I took a pass over this PR, left some comments

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: In this case, it's purely a mathematical statement.

I set the "close time" equal to the first ms in which a record is late, so by definition if window_end==close_time, it is late. Note how the comparison above uses `>`.

I could move the "close time" forward 1ms (adding 1 to the calculation of it above), used `>=` in the comparison, and then what you say would be true, and we'd have no need to add 1 here.

I think either way, we're adding a 1 to some calculation in this method, and we'd need a similar explanatory comment to clarify why.

I'm open to swapping the semantics if you think it's more intuitive the other way.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: I think the average lateness is actually one of the more important metrics for determining what timeout to use for suppression.

If you're worried about the performance, I can tackle KAFKA-7275 sooner rather than later.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: I was thinking of it like, "the hardware store closes at 7pm, so if I show up at 7pm, I can't go in". I.e., `7:00:00.000pm` is the first ms in which I can't get into the store.

Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: My thinking was if I walk into a meeting scheduled to start at `1:00:00` precisely at `1:00:00` I'm not late, but again I don't have a firm opinion here, and I can see your point as well.

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: But in this case, we're not talking about the start of the window's lifespan, we're talking about its end.

So the analogy would be more like saying "the meeting starts at `1:00:00` and ends `1hr` later." If you arrive at `1:59:59`, you'll find the meeting wrapping up, but if you arrive at `2:00:00`, you are late, and the meeting is over.

Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: Yep, I knew that, but I keep mixing up my examples; maybe I have window dyslexia.

Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: Nah, it's confusing... I'm always telling myself these little stories to try and keep it straight.

**Comment by vvcephei** [comment]
Thanks for the review, @bbejeck !

Java 10 passed, but the java 8 test failed with the same problem with the core integration tests as other builds have been.

Retest this, please.

**Comment by vvcephei** [comment]
@mjsax or @guozhangwang , would you mind reviewing this if you have a chance?

Also, it would need to be cherry-picked to 2.1

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: Note that for `TimeWindow` as we stated in javadoc ` its end timestamp as exclusive boundary` the current logic that `mergedWindow.end() <= closeTime` is considered late looks good to me.

Regarding whether we should use `closeTime - mergedWindow.end() + 1` or just `closeTime - mergedWindow.end()`, I admit the 1ms seems a bit unnatural but from monitoring purposes I think it may be better aligned with `lateRecordDropSensor`: think, if a users saw `lateRecordDropSensor` is non-zero but `recordLatenessSensor` is 0.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: 1) Current `lateRecordDropSensor` seems to be completely paired with `recordLatenessSensor`, i.e. there is no place where we only record one of them but not both. It makes me thinking if we can optimize this a bit to get one sensor, with a COUNT / and AVG / MAX metrics with it, as it seems that `lateRecordDropSensor` is just a COUNT of the `recordLatenessSensor`.

2) I'm also a bit concerned about the performance as it is a per task-processor-node-level sensor. Would better consider doing KAFKA-7275 sooner than later.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java]: Why the max / avg values would not be dependent on "how fast the test run" here as well, since MAX and AVG are actually also sampled stats?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java]: Ditto here about the fixed values, as Max / Avg are sampled stat as well.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java]: Ditto here: are we certain there is no flakiness introduced even with `isRoughly` approximation?

**Review by guozhangwang — COMMENTED** [review]
@vvcephei My main question is on the possibility of flakiness in unit tests because Max and Avg are sampled stat which may be dependent on running time, others are minor.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java]: 1. I think you've actually pointed out a bug in my measurement. I intended to measure record lateness regardless of whether it was *too late* or not. I'll fix it.

2. Ok, I'll consider KAFKA-7275 asap.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java]: I think it is, just not as sensitively. The sampled stats are dependent on the number of samples limit (default=2) and the sample age limit (default=30s iirc). As long as neither of these limits are reached during the test, these stats will be 100% accurate.

Without looking deeply into the implementation, I would think that the rate, on the other hand, would depend on exactly how many seconds elapse during the measurement period.

Regardless, you raise a good point, and I'll consider how I can make the test more deterministic.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java]: see above

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java]: see above

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java]: I reckon that one easy fix would be to configure the samples and sample age to be super high...

Or I could make the metrics fully deterministic with MockTime, but that seems a little overkill. I don't need to test that the stats themselves compute what they claim to compute. I am more attempting to verify:
1. I have in fact recorded the metrics, with the proper names
2. I recorded the correct values (i.e., I didn't use the current time in place of the latency, or a latency as a number of occurrences, etc.)
3. I *probably* selected the right stats (i.e., I didn't use a Total when I meant to use a Max, etc.)

For these purposes, I think it's sufficient to:
1. check that the metrics are present and non-zero
2. check that the values are in the right ballpark, although when it's easy to see what the exact value should be, that's nicer.

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java]: Making sample age super high sounds better to me, comparing to having a ballpark check.

Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java]: Oh, forgive me, the metric you're calling out is average lateness, not a rate.

I do expect to deterministically get the exact same value each time, for the reasons above. I was just being lazy about specifying it.

For some reason, I thought you were talking about a rate metric... I must have been thinking of the other PR I'm working on.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java:106]: So, to return to @bbejeck 's concern... This metric gets updated once for every record we process. This may be less frequent than for every record in every call to a window aggregator, but it's still pretty darn frequent.

Should I make it DEBUG-level until KAFKA-7275 is done?

**Review by vvcephei — COMMENTED** [review]
@guozhangwang @bbejeck ,

Apologies. Upon reflection, I got confused and mis-implemented the record-lateness metric, leading to some confusion in the PR.

The metric was intended to be a measurement of the lateness of all records, relative to stream time, not relative to the grace period.

I've corrected it now, in PartitionGroup.

I do think there's some merit in the concern about the performance overhead of it... Is once per record per task too frequent for an info-level sensor?

If so, I can make it debug for 2.1, then plan to tackle KAFKA-7275 and move it to info for 2.2+.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java:106]: I'm +1 on keeping both `dropped` and `lateness` sensors DEBUG for now.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java
Comment (vvcephei) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/metrics/Sensors.java:54]: This is debug now

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java:106]: The `dropped` metric was pre-existing (was `skipped-records` before). It has been at INFO level since 1.1 or 1.2. I think we should leave it at INFO. (The prior justification for this was that it should be abnormal to be dropping records, and it would probably be nice to know by default.)

I have made `lateness` a DEBUG metric, though.

File: streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/test/StreamsTestUtils.java:54]: might as well default the metric recording level to DEBUG for the tests...

File: streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java
Comment (vvcephei) [inline streams/test-utils/src/main/java/org/apache/kafka/streams/TopologyTestDriver.java:266]: I had to configure the metrics in order to verify the metric with TopologyTestDriver. It also seemed like a good idea to pass in the same mock time we use for the driver itself.

**Review by vvcephei — COMMENTED** [review]
@guozhangwang @bbejeck ,

I made the new metric DEBUG, but left the existing metric at INFO.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java
Comment (vvcephei) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java:111]: The style check failed because this method was over 150 lines long.

I've consolidated the checks into verifying the timestamps (record and system) and verifying the number of buffered records (total, partition1, partition2).

**Comment by vvcephei** [comment]
Test failures unrelated:
```
kafka.api.ClientIdQuotaTest.testThrottledProducerConsumer
kafka.api.ConsumerBounceTest.testClose
kafka.api.UserQuotaTest.testQuotaOverrideDelete
```

Retest this, please.

**Comment by guozhangwang** [comment]
Made another pass over the PR. @bbejeck could you do a second review as well?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/PartitionGroup.java:106]: makes sense to me, +1 on making `lateness` DEBUG

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java]: super nit: should we assert greater than `0.0`? But I don't have a strong opinion here though.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/PartitionGroupTest.java]: nit: method can be `private`

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java:337]: For the `shouldLogAndMeterWhenSkippingExpiredWindowByGrace` and the `shouldLogAndMeterWhenSkippingExpiredWindow`  only the first record is not dropped (I know this is due to setting the stream time) and all subsequent records are dropped.  
Do we want to put a record in the middle that isn't late and thus not dropped?

**Review by bbejeck — COMMENTED** [review]
Thanks for updating @vvcephei, overall looks good to me just a couple of minor comments.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java:337]: Just to clarify, the intent here is to test the logging and metering, not the dropping behavior. Are you suggesting that we should verify that the record in the middle does not get dropped *and therefore* it doesn't get logged or contribute to the metrics?

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregateProcessorTest.java]: I was being lazy, since we're just coincidentally pulling in some, but not all, of hamcrest, we don't have access to the "greater than" matcher.

I've added the dependency now, though, so I can add this matcher as you suggest.

**Comment by vvcephei** [comment]
Thanks for the review, @bbejeck . Just one clarifying question...

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java
Comment (bbejeck) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java:337]: Yes that was the intent of my suggestion, but considering you're clarification I think the test is fine as is.

**Review by bbejeck — APPROVED** [review]
Thanks @vvcephei, LGTM assuming test pass.

**Comment by guozhangwang** [comment]
Thanks for confirming @bbejeck .

I'll merge the PR once jenkins passed.

File: streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregateTest.java:337]: Ok, yeah, I agree. We could augment the test, but I think it's also fine as is.

**Comment by vvcephei** [comment]
java 8 passed. jdk 11 build ran out of heap space.

Retest this, please.

**Comment by bbejeck** [comment]
retest this please

**Comment by vvcephei** [comment]
jdk8 failed this time (unrelated). I've submitted #5792 to fix it.

**Comment by guozhangwang** [comment]
Cherry-picked to 2.1.

**Pull request merged**
